// fetch-curl/dist/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const status = require("http-status");
const headers_1 = require("./headers");
class Response {
    constructor(req, statusCode, body, headers) {
        const countRedirects = Number(req._curl.getInfo("REDIRECT_COUNT"));
        this.json = this.getFuncJson(body);
        this.text = this.getFuncText(body);
        this.headers = new headers_1.Headers(headers[0]);
        this.ok = statusCode >= 200 && statusCode < 300;
        this.redirected = countRedirects > 0;
        this.countRedirect = countRedirects;
        this.status = statusCode;
        this.statusText = status[statusCode];
        this.url = String(req._curl.getInfo("EFFECTIVE_URL"));
    }
    isRedirect(code) {
        return code >= 301 && code <= 308;
    }
    getFuncText(body) {
        return () => {
            if (typeof body === "string") {
                return body;
            }
            return "";
        };
    }
    getFuncJson(body) {
        return () => {
            if (typeof body === "string") {
                return body ? JSON.parse(body) : "";
            }
            return {};
        };
    }
}
exports.Response = Response;
